adev/src/app/core/services/header.service.ts

```diff
@@ -0,0 +1,32 @@
+import {DOCUMENT} from '@angular/common';
+import {Injectable, inject} from '@angular/core';
+
+const ANGULAR_DEV = 'https://angular.dev';
+
+/**
+ * Information about the deployment of this application.
+ */
+@Injectable({providedIn: 'root'})
+export class HeaderService {
+  private readonly document = inject(DOCUMENT);
+
+  /**
+   * Sets the canonical link in the header.
+   * It supposes the header link is already present in the index.html
+   *
+   * The function behave invariably and will always point to angular.dev,
+   * no matter if it's a specific version build
+   */
+  setCanonical(absolutePath: string): void {
+    const pathWithoutFragment = this.normalizePath(absolutePath).split('#')[0];
+    const fullPath = `${ANGULAR_DEV}/${pathWithoutFragment}`;
+    this.document.querySelector('link[rel=canonical]')?.setAttribute('href', fullPath);
```

Comment:
All the needed information can already be found with `location.origin` and `location.pathname`, both available on CSR/SSR.

Shouldn't it be enough to do as suggested? (this mean that you no longer need `normalizePath`)

This also rely on the real origin of the page instead of the hardcoded value, meaning that the canonical url will always match the real url, maybe that's not what you want?

Example: https://next.angular.dev/ and so on will be used for the canonical.

If that's not what you want, you can change `this.document.location.origin` to `ANGULAR_DEV`.

```suggestion
    const fullPath = this.document.location.origin + this.document.location.pathname;
    this.document.querySelector('link[rel=canonical]')?.setAttribute('href', fullPath);
```

Reply:
We're doing pre-rendering, the domain isn't known at build time afaik.